pkg/kv/kvserver/kvflowcontrol/replica_rac2/admission.go

```diff
@@ -0,0 +1,277 @@
+	n := len(p.intervals)
+	// INVARIANT: if n > 0, p.intervals[0].last >= index.
+
+	// If there is no interval containing the index, return the original
+	// priority. We need to tolerate this since this may be case C1, and this is
+	// index 9, so we ignored when sideChannelForLowPriOverride provided
+	// information for 9.
+	if n == 0 || p.intervals[0].first > index {
+		return pri
+	}
+	lowPriOverride := p.intervals[0].lowPriOverride
+	// Remove the prefix of indices <= index.
+	if p.intervals[0].last > index {
+		p.intervals[0].first = index + 1
+	} else {
+		p.intervals = p.intervals[1:]
+	}
+	if lowPriOverride {
+		return raftpb.LowPri
+	}
+	return pri
+}
+
+// waitingForAdmissionState records the indices of individual entries that are
+// waiting for admission in the AC queues. These are added after emerging as
+// MsgStorageAppend in handleRaftReady, hence we can expect monotonicity of
+// indices within a leader term, and if the term changes, we can assume that
+// non-monotonicity means raft log indices being overwritten. These
+// assumptions simplify the processing of add, since it can wipe out higher
+// indices.
+type waitingForAdmissionState struct {
+	// The indices for each priority are in increasing index order.
+	//
+	// Say the indices for a priority are 3, 6, 10. We track the individual
+	// indices since when 3 is popped, we can advance admitted for that priority
+	// to 5. When 6 is popped, admitted can advance to 9. We should never have a
+	// situation where indices are popped out of order, but we tolerate that by
+	// popping the prefix upto the index being popped.
+	waiting [raftpb.NumPriorities][]admissionEntry
+}
+
+type admissionEntry struct {
+	index      uint64
+	leaderTerm uint64
+}
+
+func (w *waitingForAdmissionState) add(leaderTerm uint64, index uint64, pri raftpb.Priority) {
+	n := len(w.waiting[pri])
+	i := n
+	// Linear scan, and all the scanned items will be removed.
+	for ; i > 0; i-- {
+		if w.waiting[pri][i-1].index < index {
+			break
+		}
+		if buildutil.CrdbTestBuild {
+			if w.waiting[pri][i-1].leaderTerm >= leaderTerm {
+				panic(errors.AssertionFailedf("overwritten entry has leaderTerm %d >= %d",
+					w.waiting[pri][i-1].leaderTerm, leaderTerm))
+			}
+		}
+	}
+	// Uncommon case: i < n.
+	if i < n {
+		w.waiting[pri] = w.waiting[pri][:i]
+	}
+	w.waiting[pri] = append(w.waiting[pri], admissionEntry{
+		index:      index,
+		leaderTerm: leaderTerm,
+	})
+}
+
+func (w *waitingForAdmissionState) remove(
+	leaderTerm uint64, index uint64, pri raftpb.Priority,
+) (admittedMayAdvance bool) {
+	// Two cases:
+	// - index is found: the leaderTerm could be stale, so we ignore in that
+	//   case. Else we admit the prefix <= index.
+	// - index is not found: the leaderTerm must be stale.
+	pos := 0
+	found := false
+	// Fast path the common case.
+	if len(w.waiting[pri]) > 0 && w.waiting[pri][0].index == index {
+		found = true
+	} else {
+		pos, found = slices.BinarySearchFunc(w.waiting[pri], index,
```

Comment:
Linear scan would do well here, and cost an amortized O(1). Whereas the binary search always costs O(log N).